pkg/kube/krt/README.md

```diff
@@ -0,0 +1,229 @@
+
+A better approach would be to lift `Pod` into a primary dependency:
+
+```go
+SimplePods := krt.NewCollection[SimplePod](func(ctx krt.HandlerContext, pod *v1.Pod) *SimplePod {
+    return &SimplePod{Name: pod.Name}
+}) // Results in a Collection[SimplePod]
+```
+
+Not only is this simpler to write, its far more efficient.
+Consumers can more efficiently query for `SimplePod`s using label selectors, filters, etc.
+Additionally, if a single `Pod` changes we only recompute one `SimplePod`.
+
+Above we have a one-to-one mapping of input and output.
+We may have one-to-many mappings, though.
+In these cases, usually its best to use a `ManyCollection`.
+Like the above examples, its *possible* to express these as normal `Collection`s, but likely inefficient.
+
+Example computing a list of all container names across all pods:
+
+```go
+ContainerNames := krt.NewManyCollection[string](func(ctx krt.HandlerContext, pod *v1.Pod) (res []string) {
+    for _, c := range pod.Spec.Containers {
+      res = append(res, c.Name)
+    }
+    return res
+}) // Results in a Collection[string]
+```
+
+Example computing a list of service endpoints, similar to the Kubernetes core endpoints controller:
+
+```go
+Endpoints := krt.NewManyCollection[Endpoint](func(ctx krt.HandlerContext, svc *v1.Service) (res []Endpoint) {
+    for _, c := range krt.Fetch(ctx, Pods, krt.FilterLabel(svc.Spec.Selector)) {
+      res = append(res, Endpoint{Service: svc.Name, Pod: pod.Name, IP: pod.status.PodIP})
+    }
+    return res
+}) // Results in a Collection[Endpoint]
+```
+
+As a rule of thumb, if your `Collection` type is a list, you most likely should be using a different type to flatten the list.
+An exception to this would be if the list represents an atomic set of items that are never queried independently;
+in these cases, however, it is probably best to wrap it in a struct.
+For example, to represent the set of containers in a pod, we may make a `type PodContainers struct { Name string, Containers []string }` and have a
+`Collection[PodContainers]` rather than a `Collection[[]string]`.
+
+In theory, other forms could be expressed such as `func(input1 I1, input2 I2) *O`.
+However, there haven't yet been use cases for these more complex forms.
+
+### Transformation constraints
+
+In order for the framework to properly handle dependencies and events, transformation functions must adhere by a few properties.
+
+Basically, Transformations must be stateless and idempotent.
+* Any querying of other `Collection`s _must_ be done through `krt.Fetch`.
+* Querying other data stores that may change is not permitted.
+* Querying external state (e.g. making HTTP calls) is not permitted.
+* Transformations _may_ be called at any time, including many times for the same inputs. Transformation functions should not make any assumptions about calling patterns.
+
+Violation of these properties will result in undefined behavior (which would likely manifest as stale data).
```

Comment:
Any ideation on preventing this other than careful reviews? 

https://golangci-lint.run/contributing/new-linters/

Could potentially use go/analysis to find calls  that accept transfhttps://golangci-lint.run/contributing/new-linters/ormation functions `Walk` into those functions and error if there are calls to methods of informers or kube client (the most common mistake, I'd imagine).